### PR TITLE
feat: AU-2352, E2E: Implemented the possibility to test swapping field values

### PR DESCRIPTION
**Pull request by Alexander-Tallqvist-Siili**
# [AU-2352](https://helsinkisolutionoffice.atlassian.net/browse/AU-2352)

## What was done

This pull request implements the possibility to test field value swapping for applications.

A "field value swap" test can, for example, test that a bank account can be changed on an application that has already been saved once before. 

A "field value swap" test can be added to an application be creating a new form variant like this:

```
const fieldSwapForm: FormDataWithRemoveOptionalProps = {
  title: 'Field swap form',
  testFieldSwap: true,
  formPages: {
    "1_hakijan_tiedot": {
      items: {},
      itemsToSwap: [
        {field: 'edit-bank-account-account-number-select', swapValue: PROFILE_INPUT_DATA.iban2},
      ]
    },
    "2_avustustiedot": {
      items: {},
      itemsToSwap: [
        {field: 'edit-hankkeen-tai-toiminnan-lyhyt-esittelyteksti', swapValue: 'The new description!'},
      ]
    },
    "4_suunniteltu_toiminta": {
      items: {},
      itemsToSwap: [
        {field: 'edit-maara-kaikkiaan', swapValue: '1234'},
      ]
    },
    "6_talous": {
      items: {},
      itemsToSwap: [
        {field: 'edit-budget-static-income-entryfees', swapValue: '5678'},
        {field: 'edit-budget-static-income-sales', swapValue: '999'}
      ]
    }
  },
  expectedErrors: {},
};
```

**Note the following things:**

1. `itemsToSwap` accepts an array of objects. Each object should contain a `field` key representing the items whose value we want to swap. The object should also contain a `swapValue`, representing the value we are swapping to.

2. The original field item must have values set under the `value` and `selector` keys in the base form. This is because the "field value swap" test uses `fillFormField` to perform the swapping.

3. The implementation does currently not support swapping values for multi-value fields.

## What has changed

- `field_swap_helpers.ts`: Implemented the field swap testing functionality.

- `test_data.ts`: Implemented `FieldSwapItemList` and `FieldSwapItem` under the `itemsToSwap` key.

- `application_data_48.ts`: Added a field swapping form variant. Also updated needed selectors.

- `/tests/forms/*`: Added the new field swapping test to each test file.

- `README.md` and `.env.example`: Added info about field swapping tests.

## Other smaller changes

- `input_helpers.ts`: Added a `clearBefore` boolean to `fillFormField` and  `fillInputField`. This was added so that the any present field value can cleared before a new one is entered.

- `validation_helpers.ts`: Added the possibility to declare multiple validation classes under the `viewPageSelectors` key. If this key is set, the form validation makes sure that a form items field value is found in all the classes listed in said key when doing validation on the "/katso" page.

- `profile_input_data.ts`: Made the values in `PROFILE_INPUT_DATA` mandatory.

## How to install

Make sure your instance is up and running on correct branch.
* `git checkout feature/AU-2352-e2e-bank-account-swap-test`
* `make fresh`
* `make drush-cr`

## How to test

- Run `make test-pw-p PROJECT=forms-48` or `make test-pw-p PROJECT=forms-48-registered` if you are in a hurry. Make sure you get an output like this from the "Field swap" test:

Screenshot 2024-04-16 at 14 01 25

[AU-2352]: https://helsinkisolutionoffice.atlassian.net/browse/AU-2352?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: e2e/utils/data/profile_input_data.ts [hunk 1]
```diff
@@ -9,15 +9,15 @@
  * "View" page.
  */
 interface ProfileInputData  {
-  iban?: string;
-  iban2?: string;
-  address?: string;
-  zipCode?: string;
-  city?: string;
-  communityOfficial?: string;
```
Comment (jiisuominen) [inline e2e/utils/data/profile_input_data.ts:17]: Whoa, opettelen PR tsekkausta PhpStormin kautta!!! Mutta siis miten noi nullit voi tosta poistaa?

Comment (Alexander-Tallqvist-Siili) [reply e2e/utils/data/profile_input_data.ts:17]: @jiisuominen Noi arvot asetetaan kaikki tossa heti alapuolella, eli ne on käytännössä aina setattuja. Tein muutoksen siksi, ettei koodieditori nillitä siitä, että esim. iban voi olla undefined:
```

/**
 * The ProfileInputData interface.
 *
 * This interface provides a structure for
 * profile data. Its intention is to set up
 * profile data that can be used both when
 * creating a profile, and when verifying
 * a profiles data on an applications
 * "View" page.
 */
interface ProfileInputData  {
  iban?: string;
  iban2?: string;
  address?: string;
  zipCode?: string;
  city?: string;
  communityOfficial?: string;
  role?: string;
  email?: string;
  phone?: string;
}

/**
 * Setup profile data. This data
 * is used when creating profiles,
 * and when verifying that the profile
 * data exists on an applications "View" page.
 */
const PROFILE_INPUT_DATA: ProfileInputData = {
  iban: 'FI3147372044000048',
  iban2: 'FI5777266988169614',
  address: 'Ahonenväylä 95',
  zipCode: '91435',
  city: 'Kuopio',
  communityOfficial: 'Marko Niemi',
  role: 'Vastuuhenkilö',
  email: 'marko.niemi78@gmail.com',
  phone: '0401234567'
}
```

File: e2e/utils/validation_helpers.ts [hunk 2]
```diff
@@ -111,7 +111,7 @@ const validateExistingProfileData = async (
 
   // Grab the hard-coded input data and filter the
   // data depending on the profile type.
-  let profileInputData: ProfileInputData = PROFILE_INPUT_DATA;
+  let profileInputData: Partial<ProfileInputData> = PROFILE_INPUT_DATA;
```
Comment (jiisuominen) [inline e2e/utils/validation_helpers.ts:114]: Aaa, tää liittyy tohon datamäärittelyyn? TIL

File: e2e/utils/data/test_data.ts [hunk 3]
```diff
@@ -66,6 +67,13 @@ type HiddenItemsList = string[];
 
 type ViewPageFormatterFunction = (param: string) => string;
 
+type FieldSwapItemList = FieldSwapItem[];
+
+type FieldSwapItem = {
+  field: string;
```
Comment (jiisuominen) [inline e2e/utils/data/test_data.ts:73]: testijeejee!

**Review by jiisuominen — APPROVED** [review]
all good

**Pull request merged**